include/envelope.h

```diff
@@ -0,0 +1,98 @@
+ *  but WITHOUT ANY WARRANTY; without even the implied warranty of
+ *  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+ *  GNU General Public License for more details.
+ *
+ *  You should have received a copy of the GNU General Public License along
+ *  with this program; if not, write to the Free Software Foundation, Inc.,
+ *  51 Franklin Street, Fifth Floor, Boston, MA 02110-1301, USA.
+ */
+
+#ifndef DOSBOX_ENVELOPE_H
+#define DOSBOX_ENVELOPE_H
+
+/*  Audio Envelope
+ *  --------------
+ *  This class applies a step-wise earned-volume envelope with a fixed
+ * expiration period. The envelope is "earned" in the sense that the edge is
+ * expanded when a sample meets or exceeds it. This helps minimize the
+ * impact of unnatural waveforms that can whipsaw wildly, such as those
+ * generated from digital machine noise or binary data.
+ *
+ *  Use
+ *  ---
+ *  1. Call Update(..) to provide the Envelope with information about the audio
+ *     stream: the frame rate (in Hz), peak possible sample magnitude (from zero
+ *     to 2^16 -1), the expansion phase duration in milliseconds that represents
+ *     the shortest possible time the envelope will be expanded from zero to
+ *     peak volume if the samples "earn" it (reasonable values are <30ms), and
+ *     the desired expiration period in seconds (reasonable values are <60s).
+ *
+ *  2. Call Process(..), passing it samples in their natural 16-bit signed
+ *     form. Note: when the envelope is fully expanded or has expired, this
+ *     function becomes the a null-call, eliminating further overhead.
+ *     To be clear - there are no runtime checks you need to perform to
+ *     determine if you should use the envelope or not.  It simply goes
+ *     dormant when done.
+ *
+ *  3. Call Reactivate() to perform another round of enveloping. Note that the
+ *     characteristics about the envelope provided in the Update() call are
+ *     retained and do not need to be provided after a reactivating.
+ *
+ *  By default, the evenlope does nothing; it needs to be Update()'d for it to
+ *  do work.
+ */
+
+#include <cstdint>
+
+class Envelope {
+public:
+	void Update(uint32_t frame_rate,
+	            uint32_t peak_magnitude,
+	            uint8_t expansion_phase_ms = 15u,
+	            uint8_t expire_after_seconds = 10u);
+	void Process(bool is_stereo, bool is_interpolated, intptr_t prev[], intptr_t next[])
+	{
+		(this->*process)(is_stereo, is_interpolated, prev, next);
+	};
+	void Reactivate();
+
+private:
+	bool ClampSample(intptr_t &sample, intptr_t next_edge);
+
+	void Apply(bool is_stereo,
+	           bool is_interpolated,
+	           intptr_t prev[],
+	           intptr_t next[]);
+
+	void Skip(bool is_stereo,
+	          bool is_interpolated,
+	          intptr_t prev[],
+	          intptr_t next[]);
+
+	typedef void (Envelope::*process_f)(bool, bool, intptr_t[], intptr_t[]);
+	process_f process = &Envelope::Apply;
```

Comment:
Suggestion: maybe use `std::function` for storing function pointers?

Reply:
Will look into this; new for me - but probably more readable. Thanks for the tip!

Reply:
Got it - this is a nicer to work with.